ddtrace/tracer/option.go

```diff
@@ -556,3 +579,43 @@ func StackFrames(n, skip uint) FinishOption {
 		cfg.SkipStackFrames = skip
 	}
 }
+
+// parseHTTPCodeRanges parses HTTP codes range pairs and invalidates invalid formats.
+func parseHTTPCodeRanges(r string) []int {
+	codes := []int{}
+	for _, code := range strings.Split(r, ",") {
+		code = strings.TrimSpace(code)
+		if code == "" {
+			continue
+		}
+		valid, _ := regexp.MatchString("\\d{3}(?:-\\d{3})*(?:,\\d{3}(?:-\\d{3})*)*", code)
+		if !valid {
+			log.Warn("Invalid range for %v", code)
+			continue
+		}
+		rg := strings.Split(code, "-")
+		if len(rg) == 1 {
+			val, _ := strconv.Atoi(rg[0])
+			codes = insertCode(codes, val)
+		} else {
+			if rg[0] > rg[1] {
+				rg[0], rg[1] = rg[1], rg[0]
+			}
+			min, _ := strconv.Atoi(rg[0])
+			max, _ := strconv.Atoi(rg[1])
+			for i := min; i <= max; i++ {
+				codes = insertCode(codes, i)
```

Comment:
This is one possible approach, but this means that we are creating a new slice of integers containing hundreds of integers in some situations. Then we'd use `sort.SearchInts` (which I assume was your plan) to figure out if an HTTP status code is in the list. This is a fair approach. I wonder if we can do better.

Would it be better if we kept only the ranges? Something like `[]struct{min, max int}`, where we would make sure to reduce the user sets by merging wherever needed and ending up with the minimum amount of ranges. Then, when checking, we would iterate over that slice and simply compare the code to `min` and `max`.

I don't know which of the two is the better solution. It depends on the complexity of the algorithm. I suspect `sort.SearchInts` uses "divide et impera" so it's likely fast, but depending on the number of entries, the second approach could be faster.

I think you should try both, and write [micro-benchmarks](https://dave.cheney.net/2013/06/30/how-to-write-benchmarks-in-go) for them. Then, compare the results using [benchcmp](https://pkg.go.dev/golang.org/x/tools/cmd/benchcmp). The way to do this is to create a test set (a larger one but yet reasonable) and run it with both algorithms. After we see the results we will know the best approach.

It's important to have good performance here because this will happen on every request in a traced application.